src/upsample.jl

```diff
@@ -1,234 +1,162 @@
-    return y
-end
-
-function get_inds_and_ws(x::T, n::Int, m::Int, dim::Int) where T <: AbstractArray
-    # Creates interpolation grid for resampling. 
-    # Creates the same grid as used in Image.jl `imresize`.
-    step = n // m
-    offset = (n + 1)//2 - step//2 - step * (m//2 - 1)
-    xq = clamp.(range(offset, step=step, length=m), 1, n)
-
-    # Creates interpolation lower and upper indices, and broadcastable weights
-    ilow = floor.(Int, xq)
-    ihigh = ceil.(Int, xq)
-    sizew = ntuple(i-> i == dim ? length(xq) : 1, ndims(x))
-    wdiff = convert(T, reshape(xq .- ilow, sizew)) # wdiff possibly lives on gpu
-    return ilow, ihigh, wdiff
+# utility function
+@inline function compute_source_index_and_lambda(
+    ratio, # 0 < ratio < 1
+    output_index,
+    input_size,
+    output_size
+)
+    real_input_index = ratio*output_index
+    input_index0 = floor(Int, real_input_index) # typecast to int was here in C++
+    offset = (input_index0 < input_size - 1) ? 1 : 0
+    input_index1 = input_index0 + offset
+    lambda1 = real_input_index - input_index0
+    lambda0 = 1 - lambda1
+    return input_index0, input_index1, lambda0, lambda1
 end
 
 """
-    adjoint_of_idx(idx::Vector{<:Integer})
-
-# Arguments
-- `idx`: a vector of indices from which you want the adjoint.
+    upsample_bilinear(x::AbstractArray{T,4}, scale::NTuple{2,Real}=(1,1); outsize::Union{Nothing,NTuple{2,Integer}}=nothing)
 
-# Outputs
--`idx_adjoint`: index that inverses the operation `x[idx]`.
+Upsamples the first 2 dimensions of the array `x` by the upsample factors stored in `scale`,
+using bilinear interpolation.
 
-# Explanation
-Determines the adjoint of the vector of indices `idx`, based on the following assumptions:
-* `idx[1] == 1`
-* `all(d in [0,1] for d in diff(idx))`
-The adjoint of `idx` can be seen as an inverse operation such that:
+The size of the output is equal to
+`(scale[1]*S1, scale[2]*S2, S3, S4)`, where `S1, S2, S3, S4 = size(x)`.
 
+Examples:
 ```julia
-x = [1, 2, 3, 4, 5]
-idx = [1, 2, 2, 3, 4, 4, 5]
-idx_adjoint = adjoint_of_idx(idx)
-@assert x[idx][idx_adjoint] == x
+upsample_bilinear(x, (2, pi)) # real scaling factors are allowed
+upsample_bilinear(x; outsize=(64,64)) # note the semicolon, outsize is a keyword argument
 ```
-The above holds as long as `idx` contains every index in `x`.
+Currently only 2d upsampling is supported.
 """
-function adjoint_of_idx(idx::Vector{Int})
-    d = trues(length(idx))
-    d[2:end] .= diff(idx)
-    idx_adjoint = findall(d)
-    return idx_adjoint
+function upsample_bilinear(x::AbstractArray{T,4}, scale::NTuple{2,Real}=(1,1); outsize::Union{Nothing,NTuple{2,Integer}}=nothing) where T
+    w,h,c,n = size(x)
+    if outsize===nothing
+        out_w = floor(Int, scale[1]*w)
+        out_h = floor(Int, scale[2]*h)
+    else
+        out_w, out_h = outsize
+    end
+    y = Array{T,4}(undef, out_w, out_h, c, n)
```

Comment:
I have not checked with this implementation, but the previous one gave errors on an array of integers. I think something like this might be the simplest way to allow that:
```suggestion
    y = Array{float(T),4}(undef, out_w, out_h, c, n)
```
although I hope it won't break anything else?

Reply:
Oh this won't allow rationals, I guess.

It also won't work at all unless `upsample_bilinear_whcn!(output::AbstractArray{T,4}, input::AbstractArray{T,4})` is relaxed not to use the same `T`.

Reply:
Yes, the current implementation only works for reals and rationals. Integers can only be supported by converting forth and back I think, which could be supported by a short wrapper. Hmm and yes, maybe it makes sense to relax the type constraint. That way one would save one allocation for converting to float type and one could directly read the ints.

Reply:
Hmmm, does it really make sense to support ints here? What should the gradients be? Also int?

Reply:
One reason is to allow doc examples like `[1 2; 3 4]` to work, for which just `float(x)` would be fine. But perhaps it's worth handling better, aren't images stored as `Array{UInt8}` fairly common? 

I think gradients should not be locked to `Int`, I would have said `float(eltype(Δ))` but not if you want rationals to work.

The reason to support rationals is to allow some pretty tests?

Reply:
No, I think rationals are useless except for exact tests. I'm currently focusing on the GPU implementation and don't have the muse to code int support now, but feel free :) But we can also add it later, no problem.

Reply:
Sure, happy to fiddle once this is merged.